jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannelState.java

```diff
@@ -19,140 +19,197 @@
 package org.eclipse.jetty.websocket.core.internal;
 
 import org.eclipse.jetty.websocket.core.CloseStatus;
-
-import java.util.concurrent.atomic.AtomicReference;
+import org.eclipse.jetty.websocket.core.Frame;
+import org.eclipse.jetty.websocket.core.OpCode;
+import org.eclipse.jetty.websocket.core.ProtocolException;
 
 /**
  * Atomic Connection State
  */
 public class WebSocketChannelState
 {
-    private static class State
+    enum State
     {
-        final String name;
-        final boolean inOpen;
-        final boolean outOpen;
-        final CloseStatus closeStatus;
-
-        State(String name, boolean inOpen, boolean outOpen, CloseStatus closeStatus)
-        {
-            this.name = name;
-            this.inOpen = inOpen;
-            this.outOpen = outOpen;
-            this.closeStatus = closeStatus;
-        }
-
-        @Override
-        public String toString()
-        {
-            return String.format("%s{i=%b o=%b c=%d}", name, inOpen, outOpen, closeStatus == null?-1:closeStatus.getCode());
-        }
+        CONNECTING,
+        CONNECTED,
+        OPEN,
+        ICLOSED,
+        OCLOSED,
+        CLOSED
     }
 
-    private static final State CONNECTING = new State("CONNECTING", false, false, null);
-    private static final State CONNECTED = new State("CONNECTED", true, true, null);
-    private static final State OPEN = new State("OPEN", true, true, null);
-
-    private AtomicReference<State> state = new AtomicReference<>(CONNECTING);
+    private State _channelState = State.CONNECTING;
+    private byte _incomingContinuation = OpCode.UNDEFINED;
+    private byte _outgoingContinuation = OpCode.UNDEFINED;
+    CloseStatus _closeStatus = null;
 
     public void onConnected()
     {
-        if (!state.compareAndSet(CONNECTING, CONNECTED))
-            throw new IllegalStateException(state.get().toString());
+        synchronized (this)
+        {
+            if (_channelState != State.CONNECTING)
+                throw new IllegalStateException(_channelState.toString());
+
+            _channelState = State.CONNECTED;
+        }
     }
 
     public void onOpen()
     {
-        if (!state.compareAndSet(CONNECTED, OPEN))
-            throw new IllegalStateException(state.get().toString());
+        synchronized (this)
+        {
+            if (_channelState != State.CONNECTED)
+                throw new IllegalStateException(_channelState.toString());
+
+            _channelState = State.OPEN;
+        }
     }
 
     @Override
     public String toString()
     {
-        return state.get().toString();
+        return _channelState.toString();
+    }
+
+
+    public State getState()
+    {
+        synchronized (this)
+        {
+            return _channelState;
+        }
     }
 
     public boolean isClosed()
     {
-        State s = state.get();
-        return !s.inOpen && !s.outOpen;
+        return getState()==State.CLOSED;
     }
 
     public boolean isInOpen()
```

Comment:
Rename the method to `isInputOpen()` otherwise seems like it means it's within some "open" state (e.g. we have `Parser.isInContent()`).